### PR TITLE
nyxt-user-systems

### LINKED ISSUE #2384: Buggy nyxt-config-file
**Describe the bug**

`nyxt-config-file` appears to always return the config file that Nyxt was launched with, regardless of if a subpath is given. It is possible this is a bug in `nfiles` but not entirely sure yet. 

**Precise recipe to reproduce the issue**

It does not really matter what the config file's are, so `touch /tmp/empty.lisp`, `nyxt -C -i /tmp/empty.lisp`, and then, going into the Lisp REPL, you can run the following:

NYXT-USER> (nyxt-config-file)
#P"/tmp/empty.lisp"
NYXT-USER> (nyxt-config-file "slynk")
#P"/tmp/empty.lisp"
NYXT-USER> (nyxt-config-file "slynk.lisp")
#P"/tmp/empty.lisp"

**Information**

```
Nyxt version: 2.1.1-2720-g232eddcf
Renderer version: GI-GTK
Operating system kernel: Linux 5.10.0-15-amd64
Lisp implementation: SBCL 2.2.5 (Dynamic space size: 3221225472)
Features: (:WEBKIT2 :WEBKIT2-2.36 :WEBKIT2-CORS-ALLOWLIST :WEBKIT2-PASTE-PLAINTEXT :WEBKIT2-TRACKING :WEBKIT2-MUTE :WEBKIT2-EMOJI :WEBKIT2-MEDIA :WEBKIT2-SANDBOXING :GTK-3-22 :GTK-3-20 :GTK-3-18 :GTK-3-16 :GTK-3-14 :GTK-3-12 :GTK-3-10 :GTK-3-8 :GTK-3-6 :GTK-3-4 :GTK :GDK-3-22 :GDK-3-20 :GDK-3-18 :GDK-3-16 :GDK-3-14 :GDK-3-12 :GDK-3-10 :GDK-3-8 :GDK-3-6 :GDK-3-4 :CAIRO-1-10 :CAIRO-1-12 :GDK-PIXBUF :GLIB-2-30 :GLIB-2-32 :GLIB-2-34 :GLIB-2-36 :GLIB-2-38 :GLIB-2-40 :GLIB-2-42 :GLIB-2-44 :GLIB-2-46 :GLIB-2-48 :GLIB-2-50 :GLIB-2-52 :GLIB-2-54 :GLIB-2-56 :GLIB-2-58 :GLIB :NYXT-UNSTABLE :NYXT-G232EDDCF :NYXT-2.1.1 :NYXT-2.1 :NYXT-2 :FSET-EXT-STRINGS :CUSTOM-HASH-TABLE-NATIVE :SWANK :PLUMP-UTF-32 :PARENSCRIPT :GLOBAL-VARS :DECLARE-TYPES :NAMED-READTABLES :LPARALLEL :21BIT-CHARS :CLOSER-MOP :CL-PPCRE-UNICODE :CL-UNICODE :CL-PPCRE :CHUNGA :FLEXI-STREAMS :BORDEAUX-THREADS :LPARALLEL.WITH-CLTL2 :LPARALLEL.WITH-CAS :LPARALLEL.WITH-STEALING-SCHEDULER :SPLIT-SEQUENCE CFFI-FEATURES:FLAT-NAMESPACE CFFI-FEATURES:X86-64 CFFI-FEATURES:UNIX :CFFI CFFI-SYS::FLAT-NAMESPACE ALEXANDRIA::SEQUENCE-EMPTYP :FAST-IO-SV :FAST-IO :CL-JSON-CLOS :CL-JSON :SBCL-USES-SB-ROTATE-BYTE CHIPZ-SYSTEM:GRAY-STREAMS :THREAD-SUPPORT :ASDF3.3 :ASDF3.2 :ASDF3.1 :ASDF3 :ASDF2 :ASDF :OS-UNIX :NON-BASE-CHARS-EXIST-P :ASDF-UNICODE :X86-64 :GENCGC :64-BIT :ANSI-CL :COMMON-LISP :ELF :IEEE-FLOATING-POINT :LINUX :LITTLE-ENDIAN :PACKAGE-LOCAL-NICKNAMES :SB-CORE-COMPRESSION :SB-LDB :SB-PACKAGE-LOCKS :SB-THREAD :SB-UNICODE :SBCL :UNIX)

ASDF version: 3.3.1
ASDF registries: (NYXT-SOURCE-REGISTRY ENVIRONMENT-SOURCE-REGISTRY)
Critical dependencies: (/home/user/Repos/nyxt-git/_build/cl-cffi-gtk/gtk/cl-cffi-gtk.asd /home/user/Repos/nyxt-git/_build/cl-gobject-introspection/cl-gobject-introspection.asd /home/user/Repos/nyxt-git/_build/cl-webkit/webkit2/cl-webkit2.asd)
```

**Output when started from a shell**

Irrelevant.

### PR DESCRIPTION
**Pull request by Ambrevar**
# Description

This replaces the poorly designed `load-after-system` with an ASDF-based API:
- `define-nyxt-user-system` and `define-nyxt-user-system-and-load` to define and load a system.  The syntax is the same as with `asdf:defsystem` with some adjustments to make it easier for newcomers to list files.  See example below.
- `load-system*` is like `asdf:load-system` but it instead of failing on missing dep, it warns and skips the load.

Example:

```lisp
(define-nyxt-user-system-and-load "nyxt/user/foosystem"
  :depends-on (foo-dep)
  :components ("foo" "sub/blah.lisp")) ; ".lisp" extension is optional so as to not trip the user.
```

Result:
- If `foo-dep` does not exist, a warning is emitted, but no error and the files are _not_ loaded.
- If `foo-dep` is found, then the files "foo.lisp" and "sub/blah.lisp" in the Nyxt user config directory are loaded.

Fixes #2384.

# Discussion

- [x] A great addition would be to enable Emacs-style `with-eval-after-load` lazy loading.  I do not know if ASDF supports this.

  One approach would be to keep the `load-system*` failures in a list, and every time `load-system*` is done, loop through the list and check if the newly loaded system satisfies the dep of some failed systems.  If so, try to load it again.

- [x]  Fix bug called from a package other than `:nyxt`.

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.  

- [x] I have pulled from master before submitting this PR
- [x] My code follows the style guidelines for Common Lisp code
  - See [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf) and [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml).
- [x] I have performed a self-review of my own code
- [ ] My code has been reviewed by at least one peer (the peer review to approve a PR counts.  The reviewer must download and test the code)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] There are no merge conflicts

### CODE REVIEW

File: source/configuration.lisp [hunk 1]
```diff
@@ -341,27 +341,98 @@ Call this function from your initialization file to re-enable the default ASDF r
...
-               (log:warn "Current ASDF registries: ~{~a~^, ~}"
-                         asdf:*default-source-registries*)
-               nil))))
-    (when (and (load-system system) file)
-      (load file))))
+
+
+;; TODO: Report compilation errors.
+
+(defclass nyxt-user-system (asdf:system) ()
```
Comment (aartaka) [inline source/configuration.lisp:348]: It now feels like a separate set of functions. Maybe move to a separate file then? Something like nyxt-user-asdf.lisp etc.

File: source/start.lisp [hunk 2]
```diff
@@ -35,21 +35,6 @@ to set and use the socket without parsing any file.  Instead, the socket can be
 set from the corresponding command line option or the NYXT_SOCKET environment
 variable.")
 
-(export-always 'nyxt-config-file)
-(defun nyxt-config-file (&optional subpath)
```
Comment (aartaka) [inline source/start.lisp:39]: This one might still be useful, but the main use-case is indeed gone now.

**Review by aartaka — COMMENTED** [review]
I don't see why delete `load-after-system` and `nyxt-config-file`, because those were nice user-facing primitives.

> * [x]  A great addition would be to enable Emacs-style `with-eval-after-load` lazy loading.  I do not know if ASDF supports this.
>   One approach would be to keep the `load-system*` failures in a list, and every time `load-system*` is done, loop through the list and check if the newly loaded system satisfies the dep of some failed systems.  If so, try to load it again.

But then, if we define everything as ASDF-reliant systems, ASDF should be able to build the dependency tree on its own, right?

**Comment by Ambrevar** [comment]
> But then, if we define everything as ASDF-reliant systems, ASDF should be able to build the dependency tree on its own, right?

Yes it does, but that's not what I meant.  In Emacs, there is `with-eval-after-load` which allows you to write code that is only going to be evaluated _after_ a package has been loaded (with `require`).  But we do not know when.

If `asdf:load-system` tries to load a system with a dep that's not already defined, it will just fail.  When the dep is finally define, the previous failure won't be automatically reloaded.

This is what I've addressed with `load-system*`.

File: source/configuration.lisp (hunk 1 above)
Comment (Ambrevar) [reply source/configuration.lisp:348]: I suppose we could.

We could also move the auto-config part to a separate file.

But then are we splitting into too many files?

**Comment by Ambrevar** [comment]
Ready to merge.

Last question: Should we rename `define-nyxt-user-system` to the shorter `define-nyxt-system`?

**Comment by jmercouris** [comment]
Yes, `define-nyxt-user-system` should be `define-nyxt-system`.

**Comment by jmercouris** [comment]
In case you are wondering about the force push, I went back and forth with the changes, eventually settling on keeping it as it is. The only suggestion I may make is adding an alias/nickname for a shorter version like `define-system`. Anyways, thank you Pierre!

**Comment by jmercouris** [comment]
We can always add new names without issue, and I like how descriptive `define-nyxt-user-system` is. We can always add a shorter alias later. The other way around, that is more confusing!

**Pull request merged**